File: src/go/linked/reverseBetween.go
package linked

type ListNode struct {
	Val  int
	Next *ListNode
}

// 修改值的方式
func reverseBetween(head ListNode, left int, right int) ListNode {
	dif := right - left
	memo := make(map[int]*ListNode, dif+1)
	// 找到位置
	lNode := &head
	for i := 1; i < left; i++ {
		lNode = lNode.Next
	}
	rNode := lNode
	for j := 0; j <= dif; j++ {
		memo[j] = rNode
		rNode = rNode.Next
	}
	// 进行翻转操作
	for i := 0; i < dif+1/2; i++ {
		t := memo[i].Val
		memo[i].Val = memo[dif-i].Val
		memo[dif-i].Val = t
	}

	// 返回
	return head
}

// 头插法
func reverseBetween1(head *ListNode, left int, right int) *ListNode {
	dumpNode := &ListNode{Val: -1, Next: head}
	pre := dumpNode
	for i := 1; i < left; i++ {
		pre = pre.Next
	}
	cur := pre.Next
	for i := 0; i < right-left; i++ {
		next := cur.Next
		cur.Next = next.Next
		next.Next = pre.Next
		pre.Next = next
	}
	return dumpNode.Next
}

File: feishu/ReverseBetween.java
import struct.ListNode.ListNode;

/**
 * @author Zexho
 * @date 2022/3/26 7:10 PM
 */
public class ReverseBetween {

    public ListNode reverseBetween(ListNode head, int left, int right) {
        // 1. 找到头、尾
        ListNode h = head;
        ListNode tempHead = null;
        if (left > 1) {
            for (int i = 1; i < left - 1; i++) {
                h = h.next;
            }
            tempHead = h;
            h = h.next;
        }
        ListNode t = h;
        for (int i = left; i < right; i++) {
            t = t.next;
        }
        ListNode tmpTail = t.next;
        t.next = null;

        // 2.翻转
        ListNode n1 = null;
        ListNode n2 = h;
        ListNode n3;
        while (n2 != null) {
            n3 = n2.next;
            n2.next = n1;
            n1 = n2;
            n2 = n3;
        }

        // 3.连接
        if (left > 1) tempHead.next = t;
        h.next = tmpTail;
        return head;
    }

}
